protocols/kad/src/behaviour.rs

```diff
@@ -931,6 +931,7 @@ where
     /// This parameter is used to call [`Behaviour::bootstrap`] periodically and automatically
     /// to ensure a healthy routing table.
     pub fn bootstrap(&mut self) -> Result<QueryId, NoKnownPeers> {
+        self.bootstrap_status.on_started();
```

Comment:
What I meant is that if `peers.is_empty()` we may not want to call `self.bootstrap_status.on_started()` at all, because we aren't performing a bootstrap.

If a periodic timer needs to be reset within the `Status` then the timer reset should probably be implemented there rather than calling `on_started()`, because no actual bootstrap is started. Or maybe I didn't understand what the original issue is?

Reply:
Sorry for the long time to reply.

I do agree a little bit with what you say. However, we encountered another issue with this on another matter. 

Before periodic / automatic bootstrap, the user could always react to a `NoKnownPeers` error because it was he that triggered the bootstrap. However, since now the bootstrap is triggered automatically, the user is never notified if a `NoKnownPeers` error occurs.

That is why in our fork we have now removed the special case where we check if our routing table is empty before triggering a bootstrap : we always trigger it. Doing so allows us to receive an `OutboundQueryProgressed` with empty stats, allowing us to learn that a bootstrap has failed. We can then react to it and do some stuffs.

If you agree with this change (removing the check for empty routing table), then I think we can close this conversation because we can call `self.bootstrap_status.on_started()` because we actually always trigger one.

What do you think ? Do you agree that there is no need to check for empty routing table before triggering a bootstrap ? (no other `kad` call does this by the way).

Reply:
Not checking whether the routing table is empty would take us back to before [this commit](https://github.com/libp2p/rust-libp2p/commit/3a96ebf57ffa9f476ebbd150c6b989e0da6baaff#diff-8e328a4ff7db0500312a26c9bfc556cec1735562c39579da36e14e95ccef3da0). IMO it is good that this check was introduced, allowing the bootstrap to fail fast.

If the check isn't performed, a new query for self is created, and the first time [`next()`](https://github.com/libp2p/rust-libp2p/blob/28dd3b21571e3de374ea373deb731f2f7abc7a74/protocols/kad/src/query/peers/closest.rs#L297-L363) is called, the [query state is immediately set](https://github.com/libp2p/rust-libp2p/blob/28dd3b21571e3de374ea373deb731f2f7abc7a74/protocols/kad/src/query/peers/closest.rs#L370-L375) to `Finished`. IMO it would be better to have an error message saying that the bootstrap failed because the routing table is empty, rather than looking for an empty `OutboundQueryProgressed`.